src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp

```diff
@@ -0,0 +1,682 @@
+    const std::string & metadata_path,
+    const std::string & start_lsn,
+    const size_t max_block_size_,
+    bool allow_minimal_ddl_,
+    bool is_postgresql_replica_database_engine_,
+    Storages storages_)
+    : log(&Poco::Logger::get("PostgreSQLReaplicaConsumer"))
+    , context(context_)
+    , replication_slot_name(replication_slot_name_)
+    , publication_name(publication_name_)
+    , metadata(metadata_path)
+    , connection(std::move(connection_))
+    , current_lsn(start_lsn)
+    , max_block_size(max_block_size_)
+    , allow_minimal_ddl(allow_minimal_ddl_)
+    , is_postgresql_replica_database_engine(is_postgresql_replica_database_engine_)
+    , storages(storages_)
+{
+    for (const auto & [table_name, storage] : storages)
+    {
+        buffers.emplace(table_name, Buffer(storage));
+    }
+}
+
+
+void PostgreSQLReplicaConsumer::Buffer::fillBuffer(StoragePtr storage)
+{
+    const auto storage_metadata = storage->getInMemoryMetadataPtr();
+    description.init(storage_metadata->getSampleBlock());
+
+    columns = description.sample_block.cloneEmptyColumns();
+    const auto & storage_columns = storage_metadata->getColumns().getAllPhysical();
+    auto insert_columns = std::make_shared<ASTExpressionList>();
+
+    assert(description.sample_block.columns() == storage_columns.size());
+    size_t idx = 0;
+
+    for (const auto & column : storage_columns)
+    {
+        if (description.types[idx].first == ExternalResultDescription::ValueType::vtArray)
+            preparePostgreSQLArrayInfo(array_info, idx, description.sample_block.getByPosition(idx).type);
+        idx++;
+
+        insert_columns->children.emplace_back(std::make_shared<ASTIdentifier>(column.name));
+    }
+
+    columnsAST = std::move(insert_columns);
+}
+
+
+void PostgreSQLReplicaConsumer::readMetadata()
+{
+    try
+    {
+        metadata.readMetadata();
+
+        if (!metadata.lsn().empty())
+        {
+            auto tx = std::make_shared<pqxx::nontransaction>(*connection->conn());
+            final_lsn = metadata.lsn();
+            final_lsn = advanceLSN(tx);
+            tx->commit();
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+void PostgreSQLReplicaConsumer::insertValue(Buffer & buffer, const std::string & value, size_t column_idx)
+{
+    const auto & sample = buffer.description.sample_block.getByPosition(column_idx);
+    bool is_nullable = buffer.description.types[column_idx].second;
+
+    if (is_nullable)
+    {
+        ColumnNullable & column_nullable = assert_cast<ColumnNullable &>(*buffer.columns[column_idx]);
+        const auto & data_type = assert_cast<const DataTypeNullable &>(*sample.type);
+
+        insertPostgreSQLValue(
+                column_nullable.getNestedColumn(), value,
+                buffer.description.types[column_idx].first, data_type.getNestedType(), buffer.array_info, column_idx);
+
+        column_nullable.getNullMapData().emplace_back(0);
```

Comment:
Supported. https://github.com/ClickHouse/ClickHouse/pull/20470/files#diff-872e4c11b6d37f7f55d46a9951559c31bde5353d8b8c737681012ccec072a72bR202. It will insert NULL if column is nullable.